File: phnrec.py
#!/usr/bin/env python3

# Given a dir of 8 kHz .wav files, convert each one into a string of IPA phones.
# By analogy with ./mkscp.py.
# ./phnrec.py $L-8khz $(nproc) $L

import sys, glob, os, re, math

USAGE='''USAGE: ./phnrec.py wav_dir num_jobs lang_dir'''

if len(sys.argv) < 4 or not sys.argv[2].isdecimal():
    print(USAGE)
    exit(1)
dummy, dirWav, num_jobs, lang = sys.argv
# todo: from dirWav and lang, strip any trailing slashes.
num_jobs = int(num_jobs)
if num_jobs < 1:
    print(USAGE)
    exit(1)
if not os.path.exists(dirWav):
    print('phnrec.py: missing directory ' + dirWav)
    exit(1)

wavfiles = os.listdir(dirWav)
ids = [ re.sub(r'.*/', '', re.sub(r'\..*', '', x)) for x in wavfiles ]
if not ids:
    print('phnrec.py: no .wav files in directory ' + dirWav)
    exit(1)

if not os.path.exists(lang):
    os.mkdir(lang)
phn_base = lang + '/phn/'
if not os.path.exists(phn_base):
    os.mkdir(phn_base)

for f in glob.glob(phn_base + '*'):
    os.remove(f)

# Split the dir into num_jobs file-lists.
# Pass each fileList to its own "phnrec -l fileList".
import subprocess
num_per_job = len(wavfiles)/num_jobs

dir = "../brno-phnrec/PhnRec"
config = "CZ" # todo: also HU and RU.
config = dir + "/PHN_" + config + "_SPDAT_LCRC_N1500"

for n in range(0, num_jobs):
    jobfilename = '%s%2.2d.txt' % (phn_base, n)
    idsfilename = '%s%2.2d.ids' % (phn_base, n)
    scrfilename = '%s%2.2d.scr' % (phn_base, n)
    with open(jobfilename, 'w') as f:
        with open(idsfilename, 'w') as g:
            for m in range(math.ceil(n*num_per_job), min(len(wavfiles), math.ceil((n+1)*num_per_job))):
                f.write('{}/{}\n'.format(dirWav, wavfiles[m]))
                g.write('{}\n'.format(ids[m]))
    # Spawn a new process in the background.
    # It writes foo.wav's transcription to foo.scr.
    print([dir+"/phnrec", "-c", config, "-l", jobfilename ])
    subprocess.Popen([dir+"/phnrec", "-c", config, "-l", jobfilename ])

# After all phnrec processes finish, run ./phnrec.rb to merge the transcriptions they made.

File: phnrec.rb
#!/usr/bin/env ruby
# encoding: utf-8

# After all phnrec's spawned by phnrec.py complete,
# merge the transcriptions they made, into $L/phn/IPA-scrips.txt.

if ARGV.length < 1 then
  STDERR.puts "USAGE: #$0 <newlangdir>" # e.g., $0 tamil
  exit 1
end
$L = ARGV[0]
# todo: check the existence of input files $L-8khz/*.rec and output dir $L/phn.

# Remove any previously created transcriptions.
`rm -rf #$L/phn/*.scr`

# Reformat each transcription.
Dir.glob("#$L-8khz/*.rec") {|f|
  uttid = File.basename(f, '.rec')
  scrip = $L + "/phn/" + uttid + ".scr"
  # Discard phnrec's timing info; keep only the phones.
  # Discard useless phones int,pau,spk.
  # Convert phones from SAMPA to IPA (*before* joining into one line,
  # because node.js javascript's readline notices input only after a newline!).
  # Hack: strip trailing _'s from phones (is that SAMPA)?
  # Hack: map unusual phones down to those in PTgen/mcasr/phones.txt.
  # Join phones into one line.
  # Store output in a tmp file "scrip" instead of a local variable, for more Ruby parallelism.
  `(echo -n #{uttid}; echo -n ' '; cut -f3 -d' ' #{f} | grep -Ev 'int|pau|spk' | ./sampa2ipa.js | tr -d _ |
  sed -e 's/ʋ/v/' -e 's/oː/o/' -e 's/eː/e/' |
  tr '\n' ' '; echo) > #{scrip} &`
  # This mangles the output: tr 'ʋ' 'v'
}

# Hack to wait for those threads.
# todo: wait until phn/*.scr has as many files as *.rec,
# or until no more ./sampa2ipa.js processes are running,
# and then a moment longer;
`sleep 3`

# Merge the transcriptions.
`cat #$L/phn/*.scr | sort > #$L/phn/IPA-scrips.txt;    rm -rf #$L/phn/*.scr`
